// @amplitude/analytics-core/lib/esm/index.js
export { AmplitudeCore } from './core-client';
export { Identify } from './identify';
export { Revenue, RevenueProperty } from './revenue';
export { Destination } from './plugins/destination';
export { IdentityEventSender } from './plugins/identity';
export { Config, RequestMetadata } from './config';
export { Logger } from './logger';
export { getGlobalScope } from './global-scope';
export { getAnalyticsConnector, setConnectorDeviceId, setConnectorUserId } from './analytics-connector';
export { isNewSession } from './session';
export { getCookieName, getOldCookieName } from './cookie-name';
export { getLanguage } from './language';
export { getQueryParams } from './query-params';
export { returnWrapper } from './utils/return-wrapper';
export { debugWrapper, getClientLogConfig, getClientStates } from './utils/debug';
export { UUID } from './utils/uuid';
export { createIdentifyEvent } from './utils/event-builder';
export { MemoryStorage } from './storage/memory';
export { CookieStorage } from './storage/cookie';
export { getStorageKey } from './storage/helpers';
export { BaseTransport } from './transports/base';
export { FetchTransport } from './transports/fetch';
export { LogLevel } from './types/loglevel';
export { AMPLITUDE_PREFIX, STORAGE_PREFIX } from './types/constants';
export { IdentifyOperation, SpecialEventType } from './types/event/event';
export { ServerZone } from './types/server-zone';
export { OfflineDisabled } from './types/offline';
export { DEFAULT_CSS_SELECTOR_ALLOWLIST, DEFAULT_DATA_ATTRIBUTE_PREFIX, DEFAULT_ACTION_CLICK_ALLOWLIST, } from './types/element-interactions';
export { Status } from './types/status';
export { NetworkEventCallback, networkObserver } from './network-observer';
