// @amplitude/analytics-browser/lib/cjs/default-tracking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAttributionTrackingConfig = exports.getPageViewTrackingConfig = exports.getNetworkTrackingConfig = exports.getElementInteractionsConfig = exports.isElementInteractionsEnabled = exports.isNetworkTrackingEnabled = exports.isSessionTrackingEnabled = exports.isPageViewTrackingEnabled = exports.isFormInteractionTrackingEnabled = exports.isFileDownloadTrackingEnabled = exports.isAttributionTrackingEnabled = void 0;
var tslib_1 = require("tslib");
/**
 * Returns autocapture[event] if it is defined
 * returns defaultValue if autocapture[event] is undefined
 * otherwise returns true
 */
var isTrackingEnabled = function (autocapture, event, defaultValue) {
    if (typeof autocapture === 'boolean') {
        return autocapture;
    }
    if (autocapture !== null && typeof autocapture === 'object') {
        if (autocapture[event] === false) {
            return false;
        }
        else if (autocapture[event] === undefined) {
            return defaultValue;
        }
    }
    return true;
};
var isAttributionTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'attribution', true);
};
exports.isAttributionTrackingEnabled = isAttributionTrackingEnabled;
var isFileDownloadTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'fileDownloads', true);
};
exports.isFileDownloadTrackingEnabled = isFileDownloadTrackingEnabled;
var isFormInteractionTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'formInteractions', true);
};
exports.isFormInteractionTrackingEnabled = isFormInteractionTrackingEnabled;
var isPageViewTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'pageViews', true);
};
exports.isPageViewTrackingEnabled = isPageViewTrackingEnabled;
var isSessionTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'sessions', true);
};
exports.isSessionTrackingEnabled = isSessionTrackingEnabled;
var isNetworkTrackingEnabled = function (autocapture) {
    return isTrackingEnabled(autocapture, 'networkTracking', false);
};
exports.isNetworkTrackingEnabled = isNetworkTrackingEnabled;
/**
 * Returns true if
 * 1. autocapture === true
 * 2. if autocapture.elementInteractions === true
 * 3. if autocapture.elementInteractions === object
 * otherwise returns false
 */
var isElementInteractionsEnabled = function (autocapture) {
    if (typeof autocapture === 'boolean') {
        return autocapture;
    }
    if (typeof autocapture === 'object' &&
        (autocapture.elementInteractions === true || typeof autocapture.elementInteractions === 'object')) {
        return true;
    }
    return false;
};
exports.isElementInteractionsEnabled = isElementInteractionsEnabled;
var getElementInteractionsConfig = function (config) {
    if ((0, exports.isElementInteractionsEnabled)(config.autocapture) &&
        typeof config.autocapture === 'object' &&
        typeof config.autocapture.elementInteractions === 'object') {
        return config.autocapture.elementInteractions;
    }
    return undefined;
};
exports.getElementInteractionsConfig = getElementInteractionsConfig;
var getNetworkTrackingConfig = function (config) {
    if ((0, exports.isNetworkTrackingEnabled)(config.autocapture) && config.networkTrackingOptions) {
        return config.networkTrackingOptions;
    }
    return;
};
exports.getNetworkTrackingConfig = getNetworkTrackingConfig;
var getPageViewTrackingConfig = function (config) {
    var trackOn = function () { return false; };
    var trackHistoryChanges = undefined;
    var eventType;
    var pageCounter = config.pageCounter;
    var isDefaultPageViewTrackingEnabled = (0, exports.isPageViewTrackingEnabled)(config.defaultTracking);
    if (isDefaultPageViewTrackingEnabled) {
        trackOn = undefined;
        eventType = undefined;
        if (config.defaultTracking &&
            typeof config.defaultTracking === 'object' &&
            config.defaultTracking.pageViews &&
            typeof config.defaultTracking.pageViews === 'object') {
            if ('trackOn' in config.defaultTracking.pageViews) {
                trackOn = config.defaultTracking.pageViews.trackOn;
            }
            if ('trackHistoryChanges' in config.defaultTracking.pageViews) {
                trackHistoryChanges = config.defaultTracking.pageViews.trackHistoryChanges;
            }
            if ('eventType' in config.defaultTracking.pageViews && config.defaultTracking.pageViews.eventType) {
                eventType = config.defaultTracking.pageViews.eventType;
            }
        }
    }
    return {
        trackOn: trackOn,
        trackHistoryChanges: trackHistoryChanges,
        eventType: eventType,
        pageCounter: pageCounter,
    };
};
exports.getPageViewTrackingConfig = getPageViewTrackingConfig;
var getAttributionTrackingConfig = function (config) {
    if ((0, exports.isAttributionTrackingEnabled)(config.defaultTracking) &&
        config.defaultTracking &&
        typeof config.defaultTracking === 'object' &&
        config.defaultTracking.attribution &&
        typeof config.defaultTracking.attribution === 'object') {
        return tslib_1.__assign({}, config.defaultTracking.attribution);
    }
    return {};
};
exports.getAttributionTrackingConfig = getAttributionTrackingConfig;
